s3transfer/copies.py

```diff
@@ -0,0 +1,304 @@
+
+        # Submit requests to upload the parts of the file.
+        part_futures = []
+        progress_callbacks = get_callbacks(transfer_future, 'progress')
+
+        for part_number in range(1, num_parts + 1):
+            extra_part_args = self._extra_upload_part_args(
+                call_args.extra_args)
+            # The part number for upload part starts at 1 while the
+            # range parameter starts at zero, so just subtract 1 off of
+            # the part number
+            extra_part_args['CopySourceRange'] = calculate_range_parameter(
+                part_size, part_number-1, num_parts, transfer_future.meta.size)
+            # Get the size of the part copy as well for the progress
+            # callbacks.
+            size = self._get_transfer_size(
+                part_size, part_number-1, num_parts, transfer_future.meta.size
+            )
+            part_futures.append(
+                self._executor.submit(
+                    CopyPartTask(
+                        transfer_coordinator=transfer_coordinator,
+                        main_kwargs={
+                            'client': self._client,
+                            'copy_source': call_args.copy_source,
+                            'bucket': call_args.bucket,
+                            'key': call_args.key,
+                            'part_number': part_number,
+                            'extra_args': extra_part_args,
+                            'callbacks': progress_callbacks,
+                            'size': size
+                        },
+                        pending_main_kwargs={
+                            'upload_id': create_multipart_future
+                        }
+                    )
+                )
+            )
+
+        # Submit the request to complete the multipart upload.
+        done_callbacks = get_callbacks(transfer_future, 'done')
+        self._executor.submit(
+            CompleteMultipartUploadTask(
+                transfer_coordinator=transfer_coordinator,
+                main_kwargs={
+                    'client': self._client,
+                    'bucket': call_args.bucket,
+                    'key': call_args.key
+                },
+                pending_main_kwargs={
+                    'upload_id': create_multipart_future,
+                    'parts': part_futures
+                },
+                done_callbacks=done_callbacks,
+                is_final=True
+            )
+        )
+
+    def _get_head_object_request_from_copy_source(self, copy_source):
+        if isinstance(copy_source, dict):
+            return copy.copy(copy_source)
+        elif isinstance(copy_source, str):
+            head_object_request = {}
+            bucket, key = copy_source.split('/', 1)
+            head_object_request['Bucket'] = bucket
+            head_object_request['Key'] = key
+            result = VERSION_ID_SUFFIX.search(key)
```

Comment:
Yeah I can just import the regex constant from botocore. I think I did not import it directly just incase the constant ever got moved out of handlers.py. What do you think? It could possibly be an issue since we have a major version range dependency on botocore and handlers is something we have changed (i.e. move a method in handlers.py to utils.py)